File: common/src/main/java/me/wyne/wutils/common/ConfigUtils.java
package me.wyne.wutils.common;

import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.util.Vector;

import java.util.Collection;
import java.util.Collections;
import java.util.List;

public final class ConfigUtils {

    public static final Vector ZERO_VECTOR = new Vector();

    @SuppressWarnings("DataFlowIssue")
    public static List<String> getStringList(ConfigurationSection config, String path) {
        if (!config.isList(path))
            return config.getString(path, "").isEmpty() ? Collections.emptyList() : List.of(config.getString(path));
        return config.getStringList(path);
    }

    public static String getString(ConfigurationSection config, String path, String def) {
        if (config.isList(path))
            return reduceString(config.getStringList(path));
        return config.getString(path, def);
    }

    public static String reduceString(Collection<String> stringList) {
        return stringList.stream().reduce(ConfigUtils::reduceString).orElse("");
    }

    public static String reduceString(String s1, String s2) {
        return s1 + "\n" + s2;
    }

    public static Vector getVector(String string, Vector def) {
        var args = new Args(string, ",");
        var x = args.size() < 0 ? def.getX() : Double.parseDouble(args.get(0));
        var y = args.size() < 1 ? def.getY() : Double.parseDouble(args.get(1));
        var z = args.size() < 2 ? def.getZ() : Double.parseDouble(args.get(2));
        return new Vector(x, y, z);
    }

    public static Vector getVectorOrZero(String string) {
        return getVector(string, ZERO_VECTOR);
    }

    public static Vector getVector(ConfigurationSection config, String path, Vector def) {
        return getVector(config.getString(path, ""), def);
    }

    public static Vector getVectorOrZero(ConfigurationSection config, String path) {
        return getVector(config, path, ZERO_VECTOR);
    }

    public static String toString(Vector vector) {
        return vector.getX() + "," + vector.getY() + "," + vector.getZ();
    }

}

File: commonKt/src/main/kotlin/me/wyne/wutils/common/kotlin/config/ConfigUtils.kt
package me.wyne.wutils.common.kotlin.config

import me.wyne.wutils.common.config.ConfigUtils
import me.wyne.wutils.common.duration.TimeSpan
import org.bukkit.Material
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.potion.PotionEffectType
import java.util.EnumSet
import kotlin.random.Random

const val RANGE_DELIMITER = ".."
const val COMMA_DELIMITER = ","
const val COLON_DELIMITER = ":"
const val SPACE_DELIMITER = " "

fun ConfigurationSection.getVectorOrZero(path: String) =
    ConfigUtils.getVectorOrZero(this, path)

fun ConfigurationSection.getIntComparator(path: String) =
    ConfigUtils.getIntComparator(this, path)

fun ConfigurationSection.getDoubleComparator(path: String) =
    ConfigUtils.getDoubleComparator(this, path)

fun ConfigurationSection.getIntOperation(path: String) =
    ConfigUtils.getIntOperation(this, path)

fun ConfigurationSection.getDoubleOperation(path: String) =
    ConfigUtils.getDoubleOperation(this, path)

fun ConfigurationSection.getTimeSpan(path: String) =
    ConfigUtils.getTimeSpan(this, path)

fun ConfigurationSection.getTimeSpan(path: String, def: TimeSpan) =
    ConfigUtils.getTimeSpan(this, path, def)

fun ConfigurationSection.getMillis(path: String) =
    ConfigUtils.getMillis(this, path)

fun ConfigurationSection.getMillis(path: String, def: Long) =
    ConfigUtils.getMillis(this, path, def)

fun ConfigurationSection.getTicks(path: String) =
    ConfigUtils.getTicks(this, path)

fun ConfigurationSection.getTicks(path: String, def: Long) =
    ConfigUtils.getTicks(this, path, def)

fun ConfigurationSection.getTimeSpanRange(path: String) =
    ConfigUtils.getTimeSpanRange(this, path)

fun ConfigurationSection.getVectorRange(path: String) =
    ConfigUtils.getVectorRange(this, path)

fun ConfigurationSection.getPotionTypeEnumSet(path: String): Set<PotionEffectType> =
    getStringList(path)
        .mapNotNull { runCatching { PotionEffectType.getByName(it) }.getOrNull() }
        .toSet()

fun ConfigurationSection.getMaterialEnumSet(path: String): Set<Material> =
    ConfigUtils.getMaterialEnumSet(this, path)

inline fun <reified E : Enum<E>> ConfigurationSection.getEnumSet(key: String): EnumSet<E> =
    ConfigUtils.getEnumSet(this, key, E::class.java)

inline fun <reified E : Enum<E>> ConfigurationSection.getKeyedEnumSet(key: String): EnumSet<E> =
    ConfigUtils.getKeyedEnumSet(this, key, E::class.java)

inline fun <reified E : Enum<E>> getByName(name: String?): E? =
    ConfigUtils.getByName(name, E::class.java)

inline fun <reified E : Enum<E>> getByKeyOrName(key: String?): E? =
    ConfigUtils.getByKeyOrName(key, E::class.java)

inline fun <reified E : Enum<E>> ConfigurationSection.getByKeyOrName(path: String?): E? =
    ConfigUtils.getByKeyOrName(this, path, E::class.java)

fun ConfigurationSection.getMaterial(path: String, def: Material = Material.STONE): Material =
    getString(path)?.let { Material.matchMaterial(it) } ?: def

fun ConfigurationSection.getBlockMaterial(path: String, def: Material = Material.STONE): Material =
    getMaterial(path, def)
        .also { if (!it.isBlock) throw IllegalArgumentException("Expected block at path ${"$currentPath.$path"}") }

fun ConfigurationSection.getIntRange(path: String, def: IntRange = 0..0): IntRange {
    val args = getString(path)?.split(RANGE_DELIMITER)
    args?.let {
        return IntRange(it.first().toInt(), it.last().toInt())
    }
    return def
}

typealias DoubleRange = ClosedFloatingPointRange<Double>

fun ConfigurationSection.getDoubleRange(path: String, def: DoubleRange = 0.0..0.0): DoubleRange {
    val args = getString(path)?.split(RANGE_DELIMITER)
    args?.let {
        return it.first().toDouble()..it.last().toDouble()
    }
    return def
}

fun DoubleRange.random() =
    Random.nextDouble(start, endInclusive)